folly/experimental/exception_tracer/StackTrace.h

```diff
@@ -32,11 +34,37 @@ struct StackTrace {
   uintptr_t addresses[kMaxFrames];
 };
 
-// note: no constructor so this can be __thread.
-// A StackTraceStack MUST be placed in zero-initialized memory.
-class StackTraceStack {
+namespace detail {
+template<bool IsDebug = false>
```

Comment:
Here is how I might structure it.

```
template <size_t, bool IsDebug = false>
class StackTraceStackBaseGuard {
  uintptr_t guard() const {
    return 0;
  }
};
template <size_t I>
class StackTraceStackBaseGuard<I, true> {
 protected:
  uintptr_t guard() const {
    return guard_;
  }

 private:
  uintptr_t guard_;
};

// ...

class StackTraceStack
    : public detail::StackTraceStackBaseGuard<1, kIsDebug>,
      public detail::StackTraceStackBase,
      public detail::StackTraceStackBaseGuard<2, kIsDebug> {
  void checkGuard() const {
    assert(detail::StackTraceStackBaseGuard<1, kIsDebug>::guard() == 0);
    assert(detail::StackTraceStackBaseGuard<2, kIsDebug>::guard() == 0);
  }
};
```

Reply:
A non-type template for the particular guard is a better approach, but I ended up closer to your other suggestion below which is a lot simpler.